### ISSUE TITLE
Deploying to gh-pages overwrites history

### ISSUE THREAD

**Issue by ghost** [travis-ci/travis-ci#7562]
I followed the instructions from [here](https://docs.travis-ci.com/user/deployment/pages/):

- Add minimal config to `.travis.yml`

- Create Personal Access Token and add via Travis dashboard

- Make gh-pages no longer a protected branch because the deploy would fail (first red flag)

- Travis CI force pushes and overwrote gh-pages history

Expected result: history is preserved, deploy shows up as an additional commit
Actual result: history was rewritten to only contain the deploy commit

I had to do a `git push --force` myself to restore history. Why does the bot do a force push and not pull the target branch's history first? Is this intended behavior or am I missing a step?

**Comment by BanzaiMan** [comment]
I asked the implementer of the Pages provider @webknjaz: https://github.com/travis-ci/dpl/pull/501#discussion_r81965464.

**Comment by remyabel** [comment]
Ah thanks for that. I see from the source code it clobbers the original repository and does a force push. At the very least I'd like to see this behavior reflected in the docs or at least a link to the deploy pattern that committer is talking about.

**Comment by webknjaz** [comment]
Hi @remyabel,

Back at the time of implementing this feature I only saw articles and deployment examples doing force push always. I haven't met a use case for doing this incrementally. That is why I kept it this way.

About a month ago a friend asked me to do the same: implement possibility of preserving the history. I've started work in [`feature/enable-gh-pages-keep-history` branch](https://github.com/webknjaz/dpl/tree/feature/enable-gh-pages-keep-history), but currently I'm stuck in debugging of some weird race condition and cannot proceed with a new PR. Hopefully I'll find someone who would help me to sort out this issue and I'll be able to send it as a PR back to upstream.

Another milestone is having Ruby specs for this provided. I haven't submitted those in my original PR, since I don't know Ruby and couldn't cope with this that time. I cannot figure this out now as well, however @BanzaiMan asked me to add those, so I hope I'll ship them within the same PR.

I'll post a separate comment regarding proposed changes in the PR you sent me. If you want to help me to deal with this faster, please try solving one of above.

Thanks for your interest in my contribution :)

P.S. Oh.. And regarding the docs: @timwis volunteered to add them around NY and his PR travis-ci/docs-travis-ci-com#897 got merged recently (Mar 1). I think you could add more information knowing where you contribute :tada:

**Comment by timwis** [comment]
Just to be clear, it only override history on the gh-pages branch; not the source branch. The gh-pages branch is just for deployments. May I ask why you need the history on that branch anyway? I've only ever seen this done with a force push, but it sounds like @webknjaz knows an even better way

**Comment by webknjaz** [comment]
@timwis There's [a few nodejs packages](https://www.npmjs.com/search?q=gh-pages) doing this incremental way. It turned out lot's of people use them. I've already got requests to add support for `--keep-history` feature flag turning this on.

**Comment by remyabel** [comment]
@timwis I could be missing something, but this repo of yours https://github.com/timwis/jkan/commits/gh-pages shows a pretty similar use case to mine. My gh-pages being a "production" branch pulls changes from master, contains build commits and other settings/files (like no .travis.yml, proprietary fonts) that should not exist in other branches. Ideally when our private web server is running again, I will no longer need gh-pages.

Anyways I appreciate the hardwork that @webknjaz is doing. Looking at this commit for example (https://github.com/webknjaz/dpl/commit/01d6d5114a363d6ca7e83bf62c99b49202a58aaa) there is a lot more to consider than just what I had in mind.

I never pull gh-pages back into other branches so that is not a problem. But I think a full `git clone` instead of a `git pull --rebase=preserve` may cause issues when merging other branches into gh-pages. I could be wrong.

**Comment by webknjaz** [comment]
@remyabel @timwis @per1234 @iriberri

It looks I've fixed the problem, but will need to improve the code (make in cleaner) + add tests.

If you wish to volunteer to test it, feel free to [follow this instruction](https://github.com/webknjaz/dpl/pull/1#issuecomment-291487748).

**Comment by webknjaz** [comment]
Okay, there's still smth with race conditions (probably). It would be nice to have some help with ruby.

**Comment by jyasskin** [comment]
@webknjaz, could you elaborate on the race conditions you're seeing? What appears to be racing with what?

**Comment by webknjaz** [comment]
@jyasskin the problem is that sometimes I see stderr output from git command saying smth like it's not a git dir we're in. And it's flaky/randomly reproducible. This scares me: it looks like tmpdir is sometimes removed before `git push` executed.

**Comment by jyasskin** [comment]
@webknjaz I saw something like that when `git clone` failed (in my case because I was testing the case when the gh-pages branch didn't exist yet). When `git clone` fails, it removes its target directory, and if that's the same as the tmpdir, the tmpdir removal code throws an exception when it tries to delete a now-nonexistent directory. I fixed it in my patch by re-creating the directory when the clone fails.

**Comment by webknjaz** [comment]
@jyasskin I've found some old build log https://travis-ci.org/webknjaz/dpl-gh-pages/builds/204541804#L204
Do you think it might be the case?

**Comment by jyasskin** [comment]
Yep, that's the same error, and
```
Trying to clone a single branch multi-provider-gh-pages-nonexistent from existing repo...
warning: Could not find remote branch multi-provider-gh-pages-nonexistent to clone.
fatal: Remote branch multi-provider-gh-pages-nonexistent not found in upstream origin
```
is consistent with the `clone` removing the directory.

**Comment by webknjaz** [comment]
@jyasskin plz also takes a look at my PR #719

**Comment by webknjaz** [comment]
UPD: all the issues have been fixed in my PR #719, it's been polished additionally, there's a few new feature-flags to facilitate possible additional modes: token validity checks improved, added flag for allowing an empty commit, reused verbose flag for printing out internal steps.
